pkg/process/util/containers.go

```diff
@@ -11,47 +11,63 @@ import (
 
 var (
 	containerCacheDuration = 10 * time.Second
-	detector               *collectors.Detector
+	detectors              []*collectors.Detector
 )
 
-// SetContainerSource allows config to force a single container source
-func SetContainerSource(name string) {
-	detector = collectors.NewDetector(name)
+// SetContainerSources allows config to force one or multiple container sources
+func SetContainerSources(names []string) {
+	detectors = []*collectors.Detector{}
+	for _, name := range names {
+		detectors = append(detectors, collectors.NewDetector(name))
+	}
 }
 
-// GetContainers returns containers found on the machine, autodetecting
-// the best backend from available sources
+// GetContainers returns containers found on the machine
+// GetContainers autodetects the best backend from available sources
+// if the users don't specify the preferred container sources
 func GetContainers() ([]*containers.Container, error) {
-	// Detect source
-	if detector == nil {
-		detector = collectors.NewDetector("")
-	}
-	l, name, err := detector.GetPreferred()
-	if err != nil {
-		return nil, err
+	// Detect sources
+	if detectors == nil {
+		// Container sources aren't configured, autodetect the best available source
+		detectors = []*collectors.Detector{collectors.NewDetector("")}
 	}
 
-	// Get containers from cache and update metrics
-	cacheKey := cache.BuildAgentKey("containers", name)
-	cached, hit := cache.Cache.Get(cacheKey)
-	if hit {
-		containers, ok := cached.([]*containers.Container)
-		if ok {
-			err := l.UpdateMetrics(containers)
-			log.Tracef("Got %d containers from cache", len(containers))
-			return containers, err
+	result := []*containers.Container{}
+	for _, detector := range detectors {
+		l, name, err := detector.GetPreferred()
+		if err != nil {
+			return nil, err
+		}
+		// Get containers from cache and update metrics
+		cacheKey := cache.BuildAgentKey("containers", name)
+		cached, hit := cache.Cache.Get(cacheKey)
+		if hit {
+			containers, ok := cached.([]*containers.Container)
+			if ok {
+				err := l.UpdateMetrics(containers)
+				if err != nil {
+					log.Debugf("Cannot update container metrics via %s: %s", name, err)
+					continue
+				}
+				log.Tracef("Got %d containers from cache", len(containers))
+				result = append(result, containers...)
+				continue
+			}
+			log.Errorf("Invalid container list cache format, forcing a cache miss")
 		}
-		log.Errorf("Invalid container list cache format, forcing a cache miss")
-	}
 
-	// If cache empty/expired, get a new container list
-	containers, err := l.List()
-	if err != nil {
-		return nil, err
+		// If cache empty/expired, get a new container list
+		containers, err := l.List()
```

Comment:
dumb question: would there be a case that several detectors are able to detect the same containers? If so I think you'll need to dedupe.

Reply:
good point, will do 👍

Reply:
done